### ISSUE TITLE
Add message event listener when skipWaiting: false

### ISSUE THREAD

**Issue by loliconer** [vuejs/vue-cli#3704]
### What problem does this feature solve?
When developing PWA, the default `workboxPluginMode` is `generateSW`, and they set `skipWaiting: false`. In this case,  the generated service worker should allow our web app to trigger `skipWaiting()` on demand,  by using `postMessage()` to communicate with our service worker.

This feature was implemented in Workbox v4.1.0, but Vue CLI 3.x can't take advantage of it.

### What does the proposed API look like?
```js
self.addEventListener('message', (event) => {
  if (event.data && event.data.type === 'SKIP_WAITING') {
    self.skipWaiting();
  }
});
```

Reference: `skipWaiting` option in [workbox-webpack-plugin](https://developers.google.com/web/tools/workbox/modules/workbox-webpack-plugin)

**Comment by loliconer** [comment]
@LinusBorg 
Yes, I can send a message to the SW, my code in `registerServiceWorker.js`:
```js
    updated (reg) {
      console.log('New content is available; please refresh.')
      window.addEventListener('refreshSW', function () {
        reg.waiting.postMessage({ type: 'SKIP_WAITING' })
      })
    },
```
The problem is the ServiceWorker can't skip waiting after posting message, because it hasn't a listener to listen the `message` event. So the code:
```js
self.addEventListener('message', (event) => {
  if (event.data && event.data.type === 'SKIP_WAITING') {
    self.skipWaiting();
  }
});
```
should be included in the `service-worker.js`, which is auto-generated. But now the code doesn't exist in `service-worker.js`.

**Comment by LinusBorg** [comment]
Ah, yeah that's because we are still using workbox 3.

Workbox 4 is on our roadmap though (see https://github.com/vuejs/vue-cli/issues/3649), and with that update, this will just work.

**Comment by loliconer** [comment]
The roadmap is a great job, thanks for your contributions. Looking forward to v4.
During Vue CLI 3.x, my workaround is auto appending the code to the generated `service-worker.js` after every build, which seems to be a little troublesome.

**Comment by LinusBorg** [comment]
well, you can always use the `injectSW` startegy instead of the default `generateSW` and add that code to your serviceworker file.

**Comment by loliconer** [comment]
Thanks. It works well.

**Comment by wizardpisces** [comment]
> well, you can always use the `injectSW` startegy instead of the default `generateSW` and add that code to your serviceworker file.

injectSW mode would cause another problem

**Comment by wizardpisces** [comment]
> Thanks. It works well.

Hi, how did you inject that code? I kind of also haunted by this problem. Thanks

**Comment by loliconer** [comment]
@wizardpisces 
The `importScripts` statement will still be auto injected into the final `service-worker.js`. So just config the `vue.config.js` like this:
```js
pwa: {
  workboxPluginMode: 'InjectManifest',
  workboxOptions: {
    swSrc: 'src/service-worker.js'
  }
}
```

**Comment by wizardpisces** [comment]
Then how to inject generateSW Mode related configuration like `runtimeCaching
offlineGoogleAnalytics
navigateFallback
directoryIndex` , did you mannually added into service-worker.js?

**Comment by loliconer** [comment]
Sorry, I did not use these configurations. You may have a try.

**Comment by wizardpisces** [comment]
> Sorry, I did not use these configurations. You may have a try.

thanks

**Comment by vesper8** [comment]
am also haunted by this problem.. @wizardpisces would you mind sharing your code/strategy on how you are injecting the listener into the generated service-worker.js after every build?

**Comment by vesper8** [comment]
also, out of curiosity, instead of going through all this trouble, why not just set skipwaiting:true in vue.config ? If your plan is to trigger skipWaiting every time the code is updated, doesn't this achieve the same?

**Comment by loliconer** [comment]
@vesper8 Set `skipwaiting: true` in `vue.config.js` will directly update the frontend resources to the latest version without notify the visitors. But sometimes we want the visitors notice that there is an update and then they can update the app manually.

**Comment by vesper8** [comment]
@loliconer I actually tried to set skipWaiting: true in vue.config.js but I couldn't figure out how to do it. Would you mind sharing how that can be set so it updates right away as you said? I'd appreciate it! Thanks

**Comment by loliconer** [comment]
@vesper8 
```js
pwa: {
  workboxOptions: {
    skipWaiting: true
  }
}
```

**Comment by wizardpisces** [comment]
> am also haunted by this problem.. @wizardpisces would you mind sharing your code/strategy on how you are injecting the listener into the generated service-worker.js after every build?

Which mode did you use ? 
I use **GenerateSW** mode, **importScripts** option saved my life , you could easily inject code into service-worker.js after evert build.
If you choose InjectManifest mode, the problem would not exist at all

more detailed infomation you could find [here](https://juejin.im/post/5cac0c985188251ae95d1d99)

**Comment by vesper8** [comment]
@LinusBorg @loliconer @wizardpisces 

I am now using the v4 alpha which comes with Workbox 4

I have tried adding this `navigator.serviceWorker.controller.postMessage({ type: 'SKIP_WAITING'});
` inside the updated() hook inside `register-service-worker.js`

But this gives me the following error:

`Error during service worker registration: TypeError: Cannot read property 'postMessage' of null`

I have also tried adding
```
updated(reg) {
    window.addEventListener('refreshSW', function () {
      reg.waiting.postMessage({ type: 'SKIP_WAITING' })
    })
}
`
``

but this does nothing

I would really really like to know how to immediately clear the cache when new content is detected. I am so boggled that this is so damn hard to accomplish and how is this not affecting thousands of vue-cli developers? Everytime I push a new breaking change it seems like all my users run into issues with the cache not being cleared

**Comment by wizardpisces** [comment]
> @LinusBorg @loliconer @wizardpisces
> 
> I am now using the v4 alpha which comes with Workbox 4
> 
> I have tried adding this `navigator.serviceWorker.controller.postMessage({ type: 'SKIP_WAITING'}); ` inside the updated() hook inside `register-service-worker.js`
> 
> But this gives me the following error:
> 
> `Error during service worker registration: TypeError: Cannot read property 'postMessage' of null`
> 
> I have also tried adding
> 
> ```
> updated(reg) {
>     window.addEventListener('refreshSW', function () {
>       reg.waiting.postMessage({ type: 'SKIP_WAITING' })
>     })
> }
> `
> ``
> 
> but this does nothing
> 
> I would really really like to know how to immediately clear the cache when new content is detected. I am so boggled that this is so damn hard to accomplish and how is this not affecting thousands of vue-cli developers? Everytime I push a new breaking change it seems like all my users run into issues with the cache not being cleared
> ```
[this article might help you](https://medium.com/@dougallrich/give-users-control-over-app-updates-in-vue-cli-3-pwas-20453aedc1f2)

**Comment by vesper8** [comment]
thank you! I actually applied this very solution by following this tutorial yesterday https://www.blog.plint-sites.nl/adding-workbox-to-a-vue-cli-pwa/

except that I am not exactly giving users a choice, I force them to update. That's because my app is still under heavy development and too often I roll out breaking changes to the API that pretty much require people to use the new front-end or else things are broken, so allowing users to skip the update just results in the site being broken.

It seems I might have done something wrong however because the prompt to update only occurs after a user does a page refresh.. not when they are continuing to use the app and a new version has been pushed. Is that how it's supposed to work or is it supposed to detect that a new version is available while a user is using the current app?

**Comment by wizardpisces** [comment]
You could interval detect new version then update  or  give user a choice,  also you could wait until refresh or second time opening app then detect update,  it's very flexible.
It all depends on how important or urgent it is to update the app

**Comment by Dean-Christian-Armada** [comment]
> The script you posted is registered in the service worker by Workbox.
> 
> You can send a message to the Service worker like this:
> 
> ```js
> navigator.serviceWorker.controller.postMessage({ type: 'SKIP_WAITING'});
> ```
> 
> a good place for that would be in the `updated` hook inside the `register-service-worker.js` file that you will find in your project's `/src` folder.

This did not work accordingly. I was expecting that visiting the new site will automatically include my new updates

**Comment by leolux** [comment]
> 
> 
> The script you posted is registered in the service worker by Workbox.
> 
> You can send a message to the Service worker like this:
> 
> ```js
> navigator.serviceWorker.controller.postMessage({ type: 'SKIP_WAITING'});
> ```
> 
> a good place for that would be in the `updated` hook inside the `register-service-worker.js` file that you will find in your project's `/src` folder.

I can see that the postMessage gets send and that the service worker receives a event but the strange thing is that the event received in the service worker does not contain the data property. Instead the event only contains this: `{"isTrusted":true}`

Any idea why the event does not contain the expected message: {data:{type:"SKIP_WAITING"}} ?
